<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 An Introduction | A Primer for Computational Biology (2nd Edition)</title>
  <meta name="description" content="Chapter 6 An Introduction | A Primer for Computational Biology (2nd Edition)" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 An Introduction | A Primer for Computational Biology (2nd Edition)" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 An Introduction | A Primer for Computational Biology (2nd Edition)" />
  
  
  

<meta name="author" content="Shawn T. O’Neil, Matthew Peterson, Leslie Coonrod" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="elementary-data-types.html"/>
<link rel="next" href="variables-and-data.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="packages/prism/prism.js"></script>

<script>
  Prism.plugins.NormalizeWhitespace.setDefaults({
	'remove-trailing': false,
	'remove-indent': false,
	'left-trim': true,
	'right-trim': true,
	'remove-initial-line-feed': true,
	'break-lines': 80

	/*'break-lines': 80,
	'indent': 2,
	'remove-initial-line-feed': false,
	'tabs-to-spaces': 4,
	'spaces-to-tabs': 4*/
});
</script>

<link rel="stylesheet" href="packages/prism/prism.css">

<style>
.book .book-body .page-wrapper .page-inner section.normal pre {
  background: #fafafa;
}

.book .book-body .page-wrapper .page-inner section.normal pre>code {
  display: block;
}

.book .book-body .page-wrapper .page-inner section.normal pre[class*="language-"].line-numbers {
    position: relative;
    padding-left: 3em;
    counter-reset: linenumber;
}

.line-highlight {
  margin-top: 0.7em;
}

.temporary {
  background: none; /*linear-gradient(to right, rgba(100, 150, 180, .5) 70%, rgba(100, 150, 180, 0)); */
}

.book .book-body .page-wrapper .page-inner section.normal h6 {
  margin-top: 0;
  margin-bottom: 0;
}

.command-line-prompt {
  margin-right: 0.2em;
  border-right: none;
}

.callout-box,.exercises {
  box-shadow: rgba(60, 64, 67, 0.3) 0px 1px 2px 0px, rgba(60, 64, 67, 0.15) 0px 1px 3px 1px;
  border-radius: 4px;
  margin: 25px 10px 10px 10px;
  padding: 10px;
}

.book .book-body .page-wrapper .page-inner section.normal .callout-box h4 {
  margin-top: 0;
}

.book .book-body .page-wrapper .page-inner section.normal .exercises h4 {
  margin-top: 0;
  color: #960018;
}


.fig.center {
  margin: 0 auto;
  display: block;
}

.fig.left {
  display: inline-block;
  float: left;
}

.fig.right {
  display: inline-block;
  float: right;
}

.fig>p.caption {
  color: #AAAAAA;
  text-align: left;
}

</style>




<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./"><b>A Primer for Computational Biology <br />(2nd Edition)</b></a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#a-few-goals"><i class="fa fa-check"></i>A Few Goals</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#organization"><i class="fa fa-check"></i>Organization</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#availability"><i class="fa fa-check"></i>Availability</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i>Acknowledgements</a></li>
<li class="chapter" data-level="" data-path="dedication.html"><a href="dedication.html"><i class="fa fa-check"></i>Dedication</a></li>
<li class="part"><span><b>Part 1: The UNIX/Linux Command-Line</b></span></li>
<li class="chapter" data-level="1" data-path="context.html"><a href="context.html"><i class="fa fa-check"></i><b>1</b> Context</a><ul>
<li class="chapter" data-level="" data-path="context.html"><a href="context.html#command-lines-and-operating-systems"><i class="fa fa-check"></i>Command Lines and Operating Systems</a></li>
<li class="chapter" data-level="" data-path="context.html"><a href="context.html#a-brief-history"><i class="fa fa-check"></i>A Brief History</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="logging-in.html"><a href="logging-in.html"><i class="fa fa-check"></i><b>2</b> Logging In</a><ul>
<li class="chapter" data-level="" data-path="logging-in.html"><a href="logging-in.html#ssh-secure-shell"><i class="fa fa-check"></i>SSH: Secure Shell</a></li>
<li class="chapter" data-level="" data-path="logging-in.html"><a href="logging-in.html#command-line-access-with-cyverse-atmosphere"><i class="fa fa-check"></i>Command-Line Access with CyVerse Atmosphere</a></li>
<li class="chapter" data-level="" data-path="logging-in.html"><a href="logging-in.html#logging-out"><i class="fa fa-check"></i>Logging Out</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="the-command-line-and-filesystem.html"><a href="the-command-line-and-filesystem.html"><i class="fa fa-check"></i><b>3</b> The Command Line and Filesystem</a></li>
<li class="part"><span><b>Part 2: Programming in Python</b></span></li>
<li class="chapter" data-level="4" data-path="hello-world.html"><a href="hello-world.html"><i class="fa fa-check"></i><b>4</b> Hello, World</a><ul>
<li class="chapter" data-level="" data-path="hello-world.html"><a href="hello-world.html#hello-world-1"><i class="fa fa-check"></i>Hello, World</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="elementary-data-types.html"><a href="elementary-data-types.html"><i class="fa fa-check"></i><b>5</b> Elementary Data Types</a><ul>
<li class="chapter" data-level="" data-path="elementary-data-types.html"><a href="elementary-data-types.html#integers-floats-and-booleans"><i class="fa fa-check"></i>Integers, Floats, and Booleans</a></li>
<li class="chapter" data-level="" data-path="elementary-data-types.html"><a href="elementary-data-types.html#strings"><i class="fa fa-check"></i>Strings</a></li>
<li class="chapter" data-level="" data-path="elementary-data-types.html"><a href="elementary-data-types.html#immutability"><i class="fa fa-check"></i>Immutability</a></li>
</ul></li>
<li class="part"><span><b>Part 3: Programming in R</b></span></li>
<li class="chapter" data-level="6" data-path="an-introduction.html"><a href="an-introduction.html"><i class="fa fa-check"></i><b>6</b> An Introduction</a><ul>
<li class="chapter" data-level="" data-path="an-introduction.html"><a href="an-introduction.html#hello-world-2"><i class="fa fa-check"></i>Hello, World</a></li>
<li class="chapter" data-level="" data-path="an-introduction.html"><a href="an-introduction.html#rstudio"><i class="fa fa-check"></i>RStudio</a></li>
<li class="chapter" data-level="" data-path="an-introduction.html"><a href="an-introduction.html#librariespackages"><i class="fa fa-check"></i>Libraries/Packages</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="variables-and-data.html"><a href="variables-and-data.html"><i class="fa fa-check"></i><b>7</b> Variables and Data</a><ul>
<li class="chapter" data-level="" data-path="variables-and-data.html"><a href="variables-and-data.html#numerics-integers-characters-and-logicals"><i class="fa fa-check"></i>Numerics, Integers, Characters, and Logicals</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Primer for Computational Biology (2nd Edition)</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="an-introduction" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> An Introduction</h1>
<p>The R programming language has a rich history, tracing its roots to the S language originally developed for statistical computing in the mid-1970s at (where else?) Bell Laboratories. Later, the open-source R project extended the capabilities of S while incorporating features of languages like LISP and Scheme.</p>
<p>Many features of R are shared with Python: both are high-level, interpreted languages. (For a discussion of interpreted vs. compiled languages, see chapter 13, “<a href="">Hello, World</a>”.) Both languages provide a wide array of features and functions for common tasks, and both languages are buttressed by a staggering variety of additional packages for more specialized analyses. Superficially, much of their syntax is similar, though below the surface lie significant (and fascinating) differences.</p>
<p>Practically, the major difference between the two languages lies in what built-in features and functions are available, and what packages are available for download. Where Python is considered a “general purpose” language, R specializes in statistical analyses. Need to build a mixed nonlinear model for a large table of numerical values from a multifactorial experiment? R is probably the tool of choice. Need to count potential promoter motifs in a large sequence set? Python is likely a better candidate. R does support functionality for the types of string analyses covered in the section on Python (such as DNA sequence analysis and regular expressions), but these are currently easier to work with in Python. Python provides excellent data plotting through the <code>matplotlib</code> library, but R’s <code>ggplot2</code> library quickly became one of the dominant tools for data visualization since its initial release in 2005.</p>
<p>Where the analysis of biological data is concerned, both languages have grown rapidly. The <code>bioconductor</code> packages in R provide many statistical bioinformatics tools, while <code>BioPython</code> focuses on some statistical methods and many sequence-oriented methods such as multiple alignment. As of this writing, both languages appear be heading toward a common feature set: relatively recent Python packages such as <code>pandas</code>, <code>numpy</code>, <code>scipy</code>, and <code>statsmodels</code> add functionality that has been present in R for decades, while R has grown in general functionality and popularity. For now, though, both languages still make great additions to a computational biologist’s repertoire, and both are supported by large and engaged communities.</p>
<p>So which of these two languages (and of course Python and R are far from the only two choices) should an aspiring computational biologist learn first? Well, the placement of Python in this book is no accident. For most users, Python is a better “introductory programming experience,” even if the experience is brief, for a couple of reasons. First, much of Python was designed with education and ease of use in mind, easing the transition to computational thinking and explaining its current popularity in Computer Science departments around the world. Second, Python shares more similarity with other “mainstream” languages like Java, C, and C++ than does R, easing transference of concepts should one wish to continue on the programming journey. Further, R contains a much larger menagerie of data types and specialized syntax for working with them, as well as multiple frameworks for things like variable assignment and object orientation. Effective R programmers arguably have more to keep in mind as they work.</p>
<p>R is a remarkably flexible language. With so much flexibility comes both power and interesting ways of thinking about programming. While Python emphasizes the use of for-loops and if-statements to control program flow, R provides an alternative syntax for manipulation of data through sophisticated logical statements. (For-loops and if-statements are discussed late in this section.) Functions are quite important in Python, but in R they take on such significance that we are required to think about them at a higher level (as types of data that can be operated on by other functions). For many of the statistical tasks in which R excels, the underlying interpreter code is highly optimized or parallelized so that analyses of millions or billions of data points can be completed quickly. Finally, many excellent packages are available only for R.</p>
<p>Ultimately, though, the answer to “which language should I learn?” is as dynamic as “which language should I use?” There are good arguments to be made for (and against) all tools, and the types of skills you wish to acquire and situational needs will play a large role at any given time. Some advice: eventually, <em>learn to program in multiple languages</em>. The benefits of learning more than one language are easily on par with learning to program in the first place!</p>
<div id="hello-world-2" class="section level3 unnumbered">
<h3>Hello, World</h3>
<p>R is an <a href="">interpreted</a> language, meaning that an R program is a text file (or multiple text files, in some cases) with commands that are interpreted by another program interacting with the CPU and RAM through the operating system. On the command line, the R interpreter is simply <code>R</code>, which we can run and send commands to one at a time.</p>
<pre id=block_7059ec
     class="language-txt 
            line-numbers 
            linkable-line-numbers">
<code>
oneils@atmosphere ~$ <b>R</b>
R version 4.0.3 (2020-10-10) -- "Bunny-Wunnies Freak Out"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin17.0 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> <b>print("Hello world!")</b>
[1] "Hello world!"
>
</code></pre>
<p>When we are done working with the R interpreter this way, we can run <code>quit(save = "no")</code> to exit, instructing that any temporary data that we haven’t already explicitly saved should not be saved.</p>
<pre id=block_a63c31
     class="language-txt 
            line-numbers 
            linkable-line-numbers">
<code>
> <b>quit(save = "no")</b>
oneils@atmosphere ~$ 
</code></pre>
<p>We will occasionally need to work with the R interpreter in this way, particularly when we need to install packages. For the most part, however, we will run R programs as executable <a href="">scripts</a>, much like we did for Python. In this case, we use the <code>Rscript</code> interpreter via the familiar <code>#!/usr/bin/env Rscript</code> line, which as always must be the first line of the file. (See chapter 5, “<a href="">Permissions and Executables</a>,” for more information on creating executable script files on the command line.)</p>
<div class="fig center" style="width: 80%">
<p><img src="images/part3-01-an-introduction.Rmd.images/III.1_3_r_3_rscript_nano.png" /></p>
</div>
<p>As with other script types, we can make this script executable with <code>chmod</code> and execute it.</p>
<pre id=block_663659
     class="language-txt 
            line-numbers 
            linkable-line-numbers">
<code>
oneils@atmosphere ~$ <b>chmod +x helloworld.R</b>
oneils@atmosphere ~$ <b>./helloworld.R</b>
[1] "Hello World"
oneils@atmosphere ~$ 
</code></pre>
</div>
<div id="rstudio" class="section level3 unnumbered">
<h3>RStudio</h3>
<p>Programming in R on the command line works as well as any other language, but the most common way to program in R today is using an integrated development environment (IDE) known as RStudio. Graphical IDEs like RStudio help the programmer to write code, manage the various files associated with a project, and provide hints and documentation during the programming process. Many IDEs (like Eclipse and Xcode) are complicated pieces of software in their own right. The RStudio IDE is moderately complex, but the developers have worked hard to focus on ease of use specifically for the R programming language. It is available for Windows, OS X, and Linux, at <a href="http://rstudio.com">http://rstudio.com</a>. Installing RStudio requires first installing the R interpreter from <a href="http://www.r-project.org">http://www.r-project.org</a>.</p>
<div class="fig center" style="width: 70%">
<p><img src="images/part3-01-an-introduction.Rmd.images/III.1_5_rstudio_first.png" /></p>
</div>
<p>When first opened, RStudio presents three panes. On the left is the same interface with the interpreter that we saw by running R on the command line. On the lower right is a pane presenting tabs for a file browser, a help browser, and a panel where plots can be viewed. The upper right pane shows the history of commands that the interpreter has run since it opened and the “global environment,” illustrating some information about which variables and data the interpreter currently has stored in memory.</p>
<p>None of these three panes, however, is the one we are primarily interested in! To open up the most important pane, we need to create a new “R script” file—a text file of R commands, just like the executable script on the command line. To do this, we use the button with a green plus sign.</p>
<div class="fig center" style="width: 50%">
<p><img src="images/part3-01-an-introduction.Rmd.images/III.1_6_rstudio_new_rscript.png" /></p>
</div>
<p>The new pane is an editor for our code file. Here we’ve entered three lines of code (a line like <code>#!/usr/bin/env Rstudio</code> is only necessary for running R scripts on the command line).</p>
<div class="fig center" style="width: 70%">
<p><img src="images/part3-01-an-introduction.Rmd.images/III.1_7_rstudio_hello_world.png" /></p>
</div>
<p>The file editor pane contains a number of buttons, four of which are worth discussing immediately. First, the save button (the small blue diskette) saves the file—R script files traditionally get the file extension <code>.R</code>. Farther to the right, the <code>Run</code> button sends the highlighted section of code to the interpreter window, even if that section is multiple lines (which are executed by the interpreter in sequence, as with most languages). The next button (with the loopy blue arrow) reruns the most recently run section of code, even if it is not highlighted. Finally, the Source button runs all the code in the file, just as would the <code>Rscript</code> version on the command line. The outputs of <code>Run</code> and <code>Source</code> are shown in the interpreter pane below, in black text.</p>
<p>Note that the Run button allows the programmer to execute lines of code out of their natural order—we could just as easily run lines 2, 3, and 4 in that order (by highlighting them with the mouse and clicking Run) as we could 4 followed by 3 followed by 2 (by highlighting and running each in turn). As it turns out, programs are usually sensitive to the order in which lines of code are executed! So, as much as possible, avoid the <code>Run</code> button and instead use the <code>Source</code> button. This means that sections of code will be rerun as you develop your programs. The benefit is that you can be sure that your code will run in a non-RStudio environment as well, and you will be less likely to create confusing code.<a href="#fn13" class="footnote-ref" id="fnref13"><sup>13</sup></a> For the most part, we won’t illustrate code directly in RStudio, but rather as simple text files and snippets of them.</p>
</div>
<div id="librariespackages" class="section level3 unnumbered">
<h3>Libraries/Packages</h3>
<p>While a default installation of the R interpreter includes a huge number of functions and packages, many additional libraries have been made available on CRAN (the Comprehensive R Archive Network), providing important new functions. Fortunately, installing these libraries from CRAN is easy and requires only the interactive R interpreter and an internet connection.<a href="#fn14" class="footnote-ref" id="fnref14"><sup>14</sup></a></p>
<p>As an example, we’ll install the <code>stringr</code> package, which provides additional functions for working with character-like data types (this will be the first additional package we’ll need in later chapters). To install it at the interactive R console, we need to run <code>install.packages("stringr")</code>. You may be asked whether the package should be installed in a “personal library” and whether to create such a personal library, to which you can answer <code>y</code>. You may also be prompted to select a nearby geographical location from which to download the package, also known as a “mirror.”</p>
<p>Once the package has been installed, using it in an R script is as easy as first calling <code>library("stringr")</code> or <code>library(stringr)</code>, after which the functions provided by the library are available. In this example, we’re using the <code>str_split()</code> function provided by the <code>stringr</code> package; the printed output would be <code>"Hello" "world"</code> rather than <code>"Hello world"</code>.</p>
<pre id=block_c41fda
     class="language-r 
            line-numbers 
            linkable-line-numbers">
<code>
library(stringr)

print(str_split("Hellow world", " "))
</code></pre>
<p>Note that <code>install.packages()</code> needs to be run only once per package, and should usually be done in the interactive R interpreter. The <code>library()</code> function will need to be used (once) for each library in each R script that uses it. These calls are usually collected near the top of the script.</p>
<div class="exercises">
<h4 id="exercises-3" class="unnumbered">Exercises</h4>
<ol style="list-style-type: decimal">
<li>If you are working on the command line, create an executable file interpreted by <code>Rscript</code> and have it print some information. If you prefer to try RStudio, install it and create a new R script, having it print several lines of text and using the <code>Source</code> button to run the entire script. Experiment with the “Run” and “Re-Run” buttons as well.</li>
<li>If you are working in RStudio, use the interface to create a new “Project” (represented by a directory in the file system housing data, R scripts, and other files related to an analysis project) and create several R scripts. For one of the scripts you create, try clicking the “Compile Notebook” icon (it looks like a small notepad) to create an HTML report, and save it in the project directory.</li>
<li>If you are using RStudio, try creating a new “R Markdown” file rather than an R script file. R Markdown files allow you to mix “chunks” of R code along with text documentation, which can then be “knitted” into a nicely formatted HTML report. Save the report in the project directory.</li>
<li>Install the <code>stringr</code> library via the interactive console and write a script that uses the <code>str_split()</code> function in it. If you are using RStudio, libraries can also be installed in the “Packages” tab of the RStudio interface.</li>
</ol>
</div>

</div>
</div>
<div class="footnotes">
<hr />
<ol start="13">
<li id="fn13"><p>On the other hand, some sections of code might run for many minutes or even hours, so you could consider carefully avoiding rerunning those sections of code when needed.<a href="an-introduction.html#fnref13" class="footnote-back">↩︎</a></p></li>
<li id="fn14"><p>In R, the terms “library” and “package” are frequently used synonymously, but technically they are distinct. The library is the directory where packages (collections of code and data providing the functionality) are stored.<a href="an-introduction.html#fnref14" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="elementary-data-types.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="variables-and-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/oneilsh/apcb_bookdown/edit/master/part3-01-an-introduction.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": {},
"toc": {
"collapse": "section"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
